CvGameCoreDLL_Expansion2/CvUnit.cpp

```diff
@@ -31260,73 +31259,160 @@ int CvUnit::GetWithdrawChance(const CvUnit& attacker, const bool bCheckChances)
+	{
+		return getHasWithdrawnThisTurn() ? 0 : 100;
+	}
+	else
 	{
-		int iWithdrawChance = getExtraWithdrawal();
 		// Does attacker have a greater speed than defender? Reduce withdrawal chance for each point the attacker is faster
 		int iDefenderMovementRange = baseMoves(isEmbarked());
 		int iAttackerMovementRange = attacker.baseMoves(attacker.isEmbarked());
 		if (iAttackerMovementRange > iDefenderMovementRange)
 			iWithdrawChance += (/*-20*/ GD_INT_GET(WITHDRAW_MOD_ENEMY_MOVES) * (iAttackerMovementRange - iDefenderMovementRange));
 
-		iWithdrawChance += (/*-20*/ GD_INT_GET(WITHDRAW_MOD_BLOCKED_TILE) * iBlockedHexes);
+		iWithdrawChance += (/*-20*/ GD_INT_GET(WITHDRAW_MOD_BLOCKED_TILE) * (3 - GetNumFallBackPlotsAvailable(attacker)));
 		return iWithdrawChance;
 	}
-	else
-		return 100;
 }
 
+// returns true if the unit should withdraw from a melee attack
+bool CvUnit::CheckWithdrawal(const CvUnit& attacker) const
+{
+	VALIDATE_OBJECT
+	int iWithdrawChance = GetWithdrawChance(attacker);
+	if (iWithdrawChance == 0)
+		return false;
+
+	//include damage so the result changes for each attack
+	int iRoll = GC.getGame().getSmallFakeRandNum(10, plot()->GetPlotIndex() + GetID() + getDamage()) * 10;
+	return iRoll < iWithdrawChance;
+}
 //	--------------------------------------------------------------------------------
-bool CvUnit::DoFallBack(const CvUnit& attacker)
+bool CvUnit::DoFallBack(const CvUnit& attacker, bool bWithdraw)
 {
 	VALIDATE_OBJECT
 
 	CvPlot* pAttackerFromPlot = attacker.plot();
 	DirectionTypes eAttackDirection = directionXY(pAttackerFromPlot, plot());
 
-	int iRightOrLeftBias = (GC.getGame().getSmallFakeRandNum(10, plot()->GetPlotIndex()+GetID()+getDamage()) < 5) ? 1 : -1;
-	int iBiases[3] = {0,-1,1};
-	
-	for(int i = 0; i < 3; i++)
-	{
-		int iMovementDirection = (NUM_DIRECTION_TYPES + eAttackDirection + (iBiases[i] * iRightOrLeftBias)) % NUM_DIRECTION_TYPES;
-		CvPlot* pDestPlot = plotDirection(getX(), getY(), (DirectionTypes) iMovementDirection);
+	CvPlot* pDestPlot = NULL;
+	if (MOD_BALANCE_VP)
+		{
+		std::vector<CvPlot*> aValidPlotList;
+		for (int i = 0; i < 3; i++)
+		{
+			int iMovementDirection = (NUM_DIRECTION_TYPES + eAttackDirection + i - 1) % NUM_DIRECTION_TYPES; // possible plots to withdraw to are the plot opposite to the attacker and the two plots next to that plot
+			CvPlot* pDirectionPlot = plotDirection(getX(), getY(), (DirectionTypes)iMovementDirection);
+
+			if (pDirectionPlot && canMoveInto(*pDirectionPlot, MOVEFLAG_DESTINATION | MOVEFLAG_NO_EMBARK) && isNativeDomain(pDirectionPlot))
+			{
+				aValidPlotList.push_back(pDirectionPlot);
+			}
+		}
 
-		if(pDestPlot && canMoveInto(*pDestPlot, MOVEFLAG_DESTINATION|MOVEFLAG_NO_EMBARK) && isNativeDomain(pDestPlot))
+		if (aValidPlotList.size() == 0)
+			return false;
+		else if (aValidPlotList.size() == 1)
+			pDestPlot = aValidPlotList[0];
+		else
 		{
+			// if there's more than one valid plot, select the plot with the fewest number of enemies
```

Comment:
Possible optimization for lines 31337-31372
```c++
// if there's more than one valid plot, select the plot with the fewest number of enemies
multimap<int, CvPlot*> aNumEnemies;
for (vector<CvPlot*>::iterator it = aValidPlotList.begin(); it != aValidPlotList.end(); ++it)
{
	aNumEnemies.insert( make_pair( (*it)->GetNumEnemyUnitsAdjacent(getTeam(), getDomainType()), (*it) ) );
}
// remove plots that have more enemies than the minimum
multimap<int, CvPlot*>::iterator itRemove = aNumEnemies.lower_bound(aNumEnemies.begin()->first + 1);
aNumEnemies.erase(itRemove, aNumEnemies.end());
		
if (aNumEnemies.size() == 1)
	pDestPlot = aNumEnemies.begin()->second;
else
{
	// if there is still more than one possible plot, select the plot with the highest number of adjacent friendly units
	multimap<int, CvPlot*> aNumFriendlies;
	for (multimap<int, CvPlot*>::iterator it = aNumEnemies.begin(); it != aNumEnemies.end(); ++it)
	{
		aNumFriendlies.insert( make_pair( it->second->GetNumFriendlyUnitsAdjacent(getTeam(), getDomainType(), true, this), it->second ) );
	}
	// remove plots that have fewer friendlies than the maximum
	itRemove = aNumFriendlies.lower_bound(aNumFriendlies.rbegin()->first); // highest key is the first key in reverse order
	aNumFriendlies.erase(aNumFriendlies.begin(), itRemove);
			
	// make a random selection from the remaining plots
	multimap<int, CvPlot*>::iterator itChosenPlot = aNumFriendlies.begin();
	if (aNumFriendlies.size() > 1)
		advance(itChosenPlot, GC.getGame().getSmallFakeRandNum(aNumFriendlies.size(), plot()->GetPlotIndex() + GetID() + getDamage()) );
	pDestPlot = itChosenPlot->second;
}
```
Multimaps are powerful when you're selecting by key; especially in this instance, we only want the smallest or largest key, which are easily accessible. In this way we can minimize push_backs, for-loops and if statements.

I dunno, it may be too messy, but I had fun figuring it out.  Don't know if multimaps need to be std::multimap in this file.

tested using cpp.sh (with the C++98 compiler) and a dummy CvPlot class

Reply:
nice! I wasn't really happy with my implementation either

Reply:
You may have already done this, but please make sure that the code functions properly inside the civ dll itself.

Reply:
I tested it before making the commit, the code works.